api/utils/clientutils/resources.go

```diff
@@ -180,3 +168,27 @@ func CollectWithFallback[T any](ctx context.Context,
 
 	return out, nil
 }
+
+// Page is a client side utility which implements auto page size adjustment.
+func Page[T any](ctx context.Context, pageSize int, pageToken string,
+	pageFunc func(context.Context, int, string) ([]T, string, error)) ([]T, string, error) {
+	for {
+		page, nextToken, err := pageFunc(ctx, pageSize, pageToken)
+		if err != nil {
+			if trace.IsLimitExceeded(err) {
+				// Cut chunkSize in half if gRPC max message size is exceeded.
+				pageSize /= 2
+				// This is an extremely unlikely scenario, but better to cover it anyways.
+				if pageSize == 0 {
+					return nil, "", trace.Wrap(err, "resource is too large to retrieve")
```

Comment:
Maybe it makes sense to print the nextToken here to make debugging easier. But it's "unlikely" so not a blocker.